File: magicalString/magicalString.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int magicalString(int n) {
        if (n == 0)
            return 0;
        int res = 1;
        string s = "122";
        int p = 2;
        while (s.length() < n) {
            char temp = s.back() == '1' ? '2' : '1';
            s += string(s[p] - '0', temp);
            if (temp == '1')
                res += s[p] - '0';
            p++;
        }
        if (s.length() > n && s.back() == '1')
            res--;
        return res;
    }
};

int main() {
  Solution s;
  cout << s.magicalString(4) << endl;
  return 0;
}

File: magicalString/magicalString.java
class Solution {
    public int magicalString(int n) {
        if (n <= 3) return 1;
        int pSeed = 2;
        StringBuilder sb = new StringBuilder();
        sb.append("122");
        int res = 1;
        while (sb.length() != n && sb.length() != n + 1) {
            if (sb.charAt(pSeed) == '1') { 
                if (sb.charAt(sb.length() - 1) == '2') {
                    sb.append("1");
                    res++;
                } else {
                    sb.append("2");
                }
            } else {
                if (sb.charAt(sb.length() - 1) == '2') {
                    sb.append("11");
                    res += 2;
                } else {
                    sb.append("22");
                }
            }
            pSeed++;
        }
        if (sb.length() == n + 1 && sb.charAt(sb.length() - 1) == '1') res--;
        return res;
    }
}
